subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+
+[[sec:embedded-kotlin]]
+=== The `embedded-kotlin` plugin
+
+`embedded-kotlin` is built on top of `kotlin("jvm")` and does the following:
+
+* applies the `kotlin("jvm")` plugin using the same version as the Gradle embedded version
+* adds `kotlin-stdlib` and `kotlin-reflect` to the `compileOnly` dependency
+
+Use this plugin if you want to use the same Kotlin version to develop your plugins that the version used by your Gradle build. This is especially useful if you do not intend to distribute your plugins, for an example, for convention plugins.
+
+[[sec:kotlin-dsl]]
+=== The `kotlin-dsl` plugin
+
+`kotlin-dsl` is built on top of `kotlin-embedded` and does the following:
+
+* applies `"java-gradle-plugin"`
+* applies `kotlin-embedded`
+* applies the `kotlin-dsl-precompiled-script-plugins` allowing to use `build.gradle.kts` files
+* adds `gradleKotlinDsl()` to the `compileOnly` configuration
+* configures the `sam-with-receiver` Kotlin compiler plugin
+* configures the `kotlin-assignment` Kotlin compiler plugin
+* sets `apiVersion` and `languageVersion` to the Kotlin embedded version
+* adds the `-Xsam-conversions=class` compiler option
+* adds others compiler options for compatibility:
+** `-java-parameters` to support https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Parameter.html[Java 8 Parameter] class and getting method parameters through reflection
+** `-Xjvm-default=all` to add link:https://kotlinlang.org/docs/java-to-kotlin-interop.html#default-methods-in-interfaces[Default methods in interfaces]
+** `-Xjsr305=strict` for https://kotlinlang.org/docs/java-interop.html#compiler-configuration[increased null safety]
+
+
+The `kotlin-dsl` plugin is the-all-in-one plugin. By applying the `kotlin-dsl-precompiled-script-plugins` and the same options as Gradle when compiling scripts, it allows you to write plugin like you would write scripts, using `.gradle.kts` file.
+
+Like the `embedded-kotlin` plugin, it uses the same Kotlin version as your Gradle build. This is useful if you do not intend to distribute your plugins, for an example, for convention plugins. If you need to distribute your plugin, make sure to use a Kotlin version that is compatible with the Gradle version you are targetting
+
+
+[[sec:your_public_api]]
+== Plugin API compatibility
+
+In order to make your plugin API compatible with Groovy, avoid relying on Kotlin specific features so that the developer experience is as good for Groovy scripts as for Kotlin scripts. For an example the Kotlin compiler emits https://github.com/JetBrains/kotlin/blob/master/libraries/kotlinx-metadata/jvm/ReadMe.md[metadata] that only the Kotlin compiler can understand for features like default parameters or extension function. It is OK to use those to provide a better experience for Kotlin scripts as long as same functionality is accessible through other means in Groovy.
```

Comment:
🧐 
```suggestion
In order to make your plugin API compatible with Groovy, avoid relying on Kotlin specific features so that the developer experience is as good for Groovy scripts as for Kotlin scripts. 
For example, the Kotlin compiler emits https://github.com/JetBrains/kotlin/blob/master/libraries/kotlinx-metadata/jvm/ReadMe.md[metadata] that only the Kotlin compiler can understand for features like default parameters or extension function. 
It is OK to use those to provide a better experience for Kotlin scripts as long as the same functionality is accessible through other means in Groovy.
```